src/Http/Owin/src/OwinEnvironment.cs

```diff
@@ -478,4 +400,229 @@ public FeatureMap(Func<TFeature, object> getter, Func<object> defaultFactory, Ac
+        /// it's more lightweight to keep track of deleted keys.
+        /// </remarks>
+        private HashSet<string> _deletedKeys;
+
+        /// <remarks>
+        /// There are some <see cref="FeatureMap"/> entries that are context-dependent.
+        /// This dictionary is allocated per request, but does not contain as many entries as <see cref="_entries"/>.
+        /// </remarks>
+        private readonly IDictionary<string, FeatureMap> _contextDependentEntries;
+
+        public OwinEntries(HttpContext context)
+        {
+            _contextDependentEntries = new Dictionary<string, FeatureMap>
+            {
+                { OwinConstants.ResponseBody, new FeatureMap<IHttpResponseBodyFeature>(feature => feature.Stream, () => Stream.Null, (feature, value) => context.Response.Body = (Stream)value) }, // DefaultHttpResponse.Body.Set has built in logic to handle replacing the feature.
+            };
+        }
+
+        static string PortToString(int port) => port switch
+        {
+            80 => "80",
+            443 => "443",
+            _ => port.ToString(CultureInfo.InvariantCulture),
+        };
+
+        public int Count
+        {
+            get
+            {
+                if (_contextEntries is not null)
+                {
+                    return _contextEntries.Count;
+                }
+
+                return _entries.Count + _contextDependentEntries.Count - (_deletedKeys?.Count ?? 0);
+            }
+        }
+
+        public IDictionary<string, FeatureMap> GetFeatureMaps()
+        {
+            InitializeContextEntries();
+            return _contextEntries;
+        }
+
+        public bool TryGetValue(string key, out FeatureMap entry)
+        {
+            if (_contextEntries is not null)
+            {
+                return _contextEntries.TryGetValue(key, out entry);
+            }
+
+            if (_deletedKeys?.Contains(key) == true)
+            {
+                entry = null;
+                return false;
+            }
+
+            if (_entries.TryGetValue(key, out entry))
+            {
+                return true;
+            }
+
+            return _contextDependentEntries.TryGetValue(key, out entry);
+        }
+
+        public bool ContainsKey(string key)
+        {
+            if (_contextEntries is not null)
+            {
+                return _contextEntries.ContainsKey(key);
+            }
+
+            if (_deletedKeys?.Contains(key) == true)
+            {
+                return false;
+            }
+
+            return _entries.ContainsKey(key) || _contextDependentEntries.ContainsKey(key);
+        }
+
+        public bool Remove(string key)
+        {
+            if (_contextEntries is not null)
+            {
+                return _contextEntries.Remove(key);
+            }
+
+            if (_entries.ContainsKey(key) || _contextDependentEntries.ContainsKey(key))
+            {
+                _deletedKeys ??= new HashSet<string>();
```

Comment:
looks like we're not consistent but `new HashSet<string>(StringComparer.Ordinal)` is more readable?

Reply:
similar comment for all the string keyed dictionaries

Reply:
added `StringComparer.Ordinal` to hashset and dictionary creations